### ISSUE TITLE
Slicing of sparse matrices is slow

### ISSUE THREAD

**Issue by cortner** [JuliaLang/julia#9668]
Slicing of sparse matrices in CSC format is very slow:

I only tested this for a specific subclass, finite-difference or finite-element type matrices, N x N matrices with O(N) entries. Taking an O(N x N) sub matrix using slicing B = A[i, i] takes O(N^2) time. See the code below for a simple example.

This was briefly discussed on [julia-users](https://groups.google.com/forum/#!topic/julia-users/yW8f9iO4T54). 

``` {.julia}
# assemble the finite-difference laplacian
function fdlaplacian(N)
    # create a 1D laplacian and a sparse identity
    fdl1 = spdiagm((ones(N-1),-2*ones(N),ones(N-1)), [-1,0,1])
    # laplace operator on the full grid
    return kron(speye(N), fdl1) + kron(fdl1, speye(N))
end

# get the list of boundary dof-indices
function get_free(N)
    L = zeros(Int, N, N)
    L[2:N-1, 2:N-1] = 1
    return find(L)
end

# timing of assembly, slice and solve
function timings(N)
    Ifree = get_free(N)
    # timing for assembly
    t1=time(); A = fdlaplacian(N); t1 = time()-t1
    # timing for the boundary condition
    t2=time(); B = A[Ifree, Ifree]; t2 = time()-t2
    # timing for the solver
    t3=time(); lufact(B); t3 = time()-t3
    return t1, t2, t3
end

# run tests once to compile
timings(10)

# runs the tests
NN = 2.^(3:8)
TT = zeros(3, length(NN))
@printf("(All times are seconds)\n")
@printf("     N   |   assembly |   slice   |   lufact  | slice / N^4 \n")
@printf("---------|------------|-----------|-----------|-------------\n")
for n = 1:length(NN)
    t1, t2, t3 = timings(NN[n])
    @printf(" %4.1e |  %4.2e  | %4.2e  |  %4.2e |   %4.2e \n",
            NN[n], t1, t2, t3, t2 / NN[n]^4)
end

# OUTPUT
# (All times are seconds)
#      N   |   assembly |   slice   |   lufact  | slice / N^4 
# ---------|------------|-----------|-----------|-------------
#  8.0e+00 |  1.50e-05  | 9.06e-06  |  2.18e-04 |   2.21e-09 
#  1.6e+01 |  3.19e-05  | 9.51e-05  |  7.42e-04 |   1.45e-09 
#  3.2e+01 |  1.03e-04  | 1.64e-03  |  2.71e-03 |   1.57e-09 
#  6.4e+01 |  5.34e-04  | 2.84e-02  |  1.19e-02 |   1.69e-09 
#  1.3e+02 |  1.95e-03  | 4.63e-01  |  5.97e-02 |   1.72e-09 
#  2.6e+02 |  7.83e-03  | 7.28e+00  |  2.74e-01 |   1.69e-09 

# julia> versioninfo()
# Julia Version 0.3.4
# Commit 3392026* (2014-12-26 10:42 UTC)
# Platform Info:
#   System: Darwin (x86_64-apple-darwin13.4.0)
#   CPU: Intel(R) Core(TM) i7-3720QM CPU @ 2.60GHz
#   WORD_SIZE: 64
#   BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
#   LAPACK: libopenblas
#   LIBM: libopenlibm
#   LLVM: libLLVM-3.3
```

**Comment by mlubin** [comment]
If the slice defining the submatrix is unordered, then I'm not sure if there's a good way to get around the O(N^2) behavior. However, when the slices are ranges you can definitely achieve O(N log N).

**Comment by cortner** [comment]
replace `return find(L)` with `return sort(find(L))` does not change the timings (though I guess you did not claim this)

But in general I disagree, you can always sort the index-list, then slice, then permute again. I currently don't have access to Matlab so cannot test, but my memory is that this was never an issue there.

**Comment by mlubin** [comment]
Good point, I see that you can get O(N log N) by sorting the slices and then iterating through the columns and rows in the right order. This depends very much on the CSC format and the fact that we require row indices to be sorted within columns.

**Comment by ViralBShah** [comment]
It is worth checking if matlab/octave/scipy do much better than us. If they do, we should at least match up to the other systems.

One design I'd like to try out is tuple storage, with lazy indexing. It may not be much faster than the current slicing, but could be more intuitive to work with if we get it right.

**Comment by mlubin** [comment]
@ViralBShah, I don't think we need to do anything fancy here, just rewrite the sparse ``getindex`` to be smarter about sorting and picking out elements as @cortner and I discussed.

**Comment by ViralBShah** [comment]
We do have the implementations for the sorted indices - so just need to check the dispatch then.

Cc: @tanmaykm who may be able to fix this quickly, since he is the author of the current sparse indexing code.

**Comment by mlubin** [comment]
Hmm, from looking at the code now it already seems to be doing something smart. It doesn't seem like @cortner's example should be quadratic time.

**Comment by ViralBShah** [comment]
Here's the example in Matlab:
````
function runtimings()

  % run tests once to compile
  timings(10);

  % runs the tests
  NN = 2.^[3:8];
  TT = zeros(3, length(NN));
  fprintf(1, '(All times are seconds)\n')
  fprintf(1, '     N   |   assembly |   slice   |   lufact  | slice / N^4 \n')
  fprintf(1, '---------|------------|-----------|-----------|-------------\n')
  for n = 1:length(NN)
    [t1, t2, t3] = timings(NN(n));
    fprintf(1, ' %4.1e |  %4.2e  | %4.2e  |  %4.2e |   %4.2e \n', ...
           NN(n), t1, t2, t3, t2 / NN(n)^4);
  end
  
end

% assemble the finite-difference laplacian
function F = fdlaplacian(N)
    % create a 1D laplacian and a sparse identity
    fdl1 = spdiags([ones(N,1) -2*ones(N,1) ones(N,1)], [-1,0,1], N, N);
    % laplace operator on the full grid
    F = kron(speye(N), fdl1) + kron(fdl1, speye(N));
end

% get the list of boundary dof-indices
function f = get_free(N)
  L = zeros(N, N);
  L(2:N-1, 2:N-1) = 1;
  f = find(L);
end

% timing of assembly, slice and solve
function [t1,t2,t3] = timings(N)
    Ifree = get_free(N);
    % timing for assembly
    tic; A = fdlaplacian(N); t1 = toc;
    % timing for the boundary condition
    tic; B = A(Ifree, Ifree); t2 = toc;
    % timing for the solver
    tic; [l,u,p,q,r] = lu(B); t3 = toc;
end
````

**Comment by ViralBShah** [comment]
Julia timings for me:
````
     N   |   assembly |   slice   |   lufact  | slice / N^4 
---------|------------|-----------|-----------|-------------
 8.0e+00 |  1.63e-04  | 2.22e-05  |  9.38e-04 |   5.41e-09 
 1.6e+01 |  4.39e-05  | 1.81e-04  |  1.98e-03 |   2.76e-09 
 3.2e+01 |  1.66e-04  | 3.05e-03  |  5.91e-03 |   2.91e-09 
 6.4e+01 |  6.79e-04  | 5.07e-02  |  2.73e-02 |   3.02e-09 
 1.3e+02 |  3.05e-03  | 8.39e-01  |  1.22e-01 |   3.12e-09 
 2.6e+02 |  1.39e-02  | 1.25e+01  |  5.74e-01 |   2.92e-09 

````

MATLAB:
````
     N   |   assembly |   slice   |   lufact  | slice / N^4 
---------|------------|-----------|-----------|-------------
 8.0e+00 |  6.39e-04  | 1.40e-05  |  2.11e-04 |   3.42e-09 
 1.6e+01 |  7.66e-04  | 1.60e-05  |  1.22e-03 |   2.44e-10 
 3.2e+01 |  1.17e-03  | 4.90e-05  |  5.29e-03 |   4.67e-11 
 6.4e+01 |  3.04e-03  | 2.48e-04  |  2.69e-02 |   1.48e-11 
 1.3e+02 |  1.05e-02  | 1.09e-03  |  4.54e-01 |   4.05e-12 
 2.6e+02 |  4.37e-02  | 4.72e-03  |  1.26e+00 |   1.10e-12 
````

**Comment by ViralBShah** [comment]
@tanmaykm and I were reviewing this code and found that the implementation is good if the number of elements in a column of the sparse matrix and the row index vector are comparable. That is not the case here, where there are 5 nonzeros per column and almost 60,000 elements in the index vector. The fix is easy - instead of streaming through the column and index vector, use binary search in such cases.

**Comment by ViralBShah** [comment]
We need to do this carefully in all of sparsematrix.jl that uses a similar code pattern. Perhaps a good time to refactor code and introduce some macros.

**Comment by cortner** [comment]
@ViralBShah 
> One design I'd like to try out is tuple storage, with lazy indexing. It may not be much faster than the
> current slicing, but could be more intuitive to work with if we get it right.

Not sure whether this is related and this is the place to discuss it, but I have been wondering whether it would be worthwhile introducing a "flexible" sparse matrix type *for assembly only*. In my own codes, I use a very basic SparseMatrixTriplet type so the main assembly routine does not need to worry about the minutiae of the sparse matrix format. I would love to have something that is more sophisticated, in particular less memory hungry.

Sorry if this is out-of-place here, I can move this discussion to julia-users

**Comment by ViralBShah** [comment]
There are discussions on this in other issues and on the mailing list. Perhaps continue the thread if relevant or start a new one.

**Comment by tanmaykm** [comment]
Implemented binary search here: https://github.com/tanmaykm/julia/compare/sparseget

Right now I'm switching from linear to binary search based on the approx ratio of nnz per column and the size of the row vector: (`searchlinear = nI < (1000 * nnz(A) / A.n)` at line 1030). But this may not be correct in all cases. Not sure what will be the right way here.

With this I get:
````
     N   |   assembly |   slice   |   lufact  | slice / N^4
---------|------------|-----------|-----------|-------------
 8.0e+00 |  9.30e-05  | 1.00e-05  |  2.18e-04 |   2.44e-09 
 1.6e+01 |  4.79e-05  | 1.57e-04  |  8.49e-04 |   2.40e-09 
 3.2e+01 |  1.05e-04  | 2.59e-03  |  3.45e-03 |   2.47e-09 
 6.4e+01 |  3.77e-04  | 4.23e-02  |  1.63e-02 |   2.52e-09 
 1.3e+02 |  1.49e-03  | 7.31e-03  |  7.58e-02 |   2.72e-11 
 2.6e+02 |  8.79e-03  | 3.38e-02  |  3.64e-01 |   7.86e-12 
````

**Comment by ViralBShah** [comment]
It seems that we are still slower than Matlab on largest N=260 case, by a factor of 7x.

**Comment by ViralBShah** [comment]
I think the switch from binary search to linear should look at the absolute difference in the two vectors. We should perhaps just run some experiments to figure that out.

I wonder if `searchsortedfirst` is getting inlined.

**Comment by tanmaykm** [comment]
OK. Will also force inline `searchsortedfirst`.

**Comment by eschnett** [comment]
There exist search algorithms that are faster than a binary search, such as e.g. "arithmetic search". (Can't find a good online reference at the moment.) The basic idea is that, given the range `ilo..ihi`, you don't look at the middle `i = (ilo+ihi)/2`, but rather estimate where the element is located in this range, via something like `i = ilo + x * (array[ihi] - array[ilo]) / (ihi - ilo)`. (Obvious rounding and range checks omitted for clarity.) Given a reasonable distribution of values -- and that is likely for a sparse matrix -- this should converge as `log log N`, instead of `log N` as for a binary search.

In plain words: If you are looking for "Addams" in a phone book, you're not opening in the middle, but rather near the beginning.

This could be implemented in `searchsorted` instead of the current binary algorithm there.

**Comment by StefanKarpinski** [comment]
I'm not convinced (although I could be) that the asymptotic improvement from `log(n)` to `log^2(n)` is really worth the additional complexity or the worsened worst case behavior. I suspect that just making searchsorted* faster by a good constant by optimizing it will be the bigger win.

**Comment by ViralBShah** [comment]
That's the first line of attack. Having really fast `searchsorted*` will generally benefit us overall.

**Comment by ViralBShah** [comment]
@tanmaykm and I just spent a bunch of time looking at this. We will shortly have the factor of 1000 that we need. `searchsorted*` is good enough. Currently the algorithm intersects every column with the row index vector `I`, and this is really slow, when `I` is huge and each column has only a few nonzeros - quite a common case.

The trick is to use binary search there. But, that doesn't get us all the way. The other observation is that we can cache the results at low cost, since most row indices will be looked up multiple times. The algorithm will thus be:

1. Have one phase where we intersect a sorted list of all unique row indices in the sparse matrix with the row index vector. This will make the best use of the binary search giving it the smallest ranges to look for the matches in, rather than going column by column and restarting the search. Later on, one could even do a double binary search to do fast intersection of two ordered lists - http://www.cs.ucr.edu/~stelo/cpm/cpm04/25_Baeza-yates.pdf

2. Store all values in a cache.

3. Construct the result by only looking up index values in the cache in a second pass.

This should not slow down existing codes, but speed up cases like the one discussed in this example almost 1000x.

**Comment by ViralBShah** [comment]
@tanmaykm It just occured to me that if we have one pass for filling the cache, we will also have enough information to compute the space requirements.

**Comment by ViralBShah** [comment]
It turns out that the inlining of searchsorted methods does not improve performance here.

I also like this benchmark a lot, and will add it to perf.

**Comment by cortner** [comment]
thanks for fixing this so quickly - amazing.
